src/execution/sema/sema_builtin.cpp

```diff
@@ -1840,6 +1840,69 @@ void Sema::CheckMathTrigCall(ast::CallExpr *call, ast::Builtin builtin) {
   call->SetType(GetBuiltinType(return_kind));
 }
 
+void Sema::CheckAtomicCall(ast::CallExpr *call, ast::Builtin builtin) {
+  const auto &call_args = call->Arguments();
+
+  int arg_count;
```

Comment:
This was a little difficult to follow and I think can benefit from refactoring. I'd prefer collapsing all logic into one `switch`, one block for the `And` and `Or`, and another for the `CompareExchange`. Then check argument counts appropriately. Also change the size check to `IsIntegral() and Size() <= 8` for clarity (rather than the verbose checks across switches and branches - make it a lambda check if you want).

You also need to check for integer types in the compare-exchange case.

Reply:
Acknowledge all but the last.

I deliberately did not check for integer on the compare-exchange case to allow for it to be used for pointer swapping that is still type-checked.  For compare-exchange, I think swapping in pointers (specifically undo records) will likely be the most common use case and I think we will open ourselves up to unnecessary errors if we have to convert to an integral before swapping as you may end up swapping in an incompatible type.

Reply:
Compare-exchanging pointers makes sense to me. I brought it up because the PR description only listed integer types.

If you want pointers in this PR, add some tests for them, too. And add comments in the refactored function stating this.

Reply:
On second look, I don't think TPL currently supports the primary use case, atomic singly-linked list, because it does not support recursive data structures at this time.  I've inserted a TODO and implemented the integral enforcement for `CompareExchange`.